python/vowpalwabbit/DFtoVWtoTensorboard.py

```diff
@@ -0,0 +1,208 @@
+from vowpalwabbit import pyvw
+from datetime import datetime
+import tensorboardX as tx
+
+
+class DFtoVWtoTensorboard:
+	"""This class provides capability for learning possible on DFtoVW object, provides capability to 
+	print metrics (average_loss, since_last, label, prediction, num_features) and also provides support 
+	to log these metrics for Tensorboard Visualization"""
+
+	def __init__(self, df_to_vw, vw):
+		"""Construct a DFtoVWtoTensorboard object
+
+		Parameters
+		----------
+
+		df_to_vw : A DFtoVW object
+					This would be used to construct examples 
+		vw		 : A vowpalwabbit object
+ 					This would be used for learning using the algorithm specified in the object, and also to compute a few metrics
+
+		Returns
+		-------
+
+		self : DFtoVWtoTensorboard
+		"""
+		self.df_to_vw = df_to_vw
+		self.vw = vw
+		self.vw_formatted_data = df_to_vw.convert_df()
+
+
+	#------------------------------------------------------------------------------------------------		
+	def _calculate_average_loss(self, sum_loss, weighted_examples):
+		"""Calculate average loss
+
+		Parameters
+		----------
+
+		sum_loss			: float
+								This is the sum loss which is used to calculate average_loss
+		weighted_examples   : float
+								This is the weighted_example which is used to calculate average_loss
+
+		Returns
+		-------
+
+		average_loss      : float
+								average_loss calculated by sum_loss and weigted_examples or if ZeroDivisionError is encountered then 0. is returned 
+		""" 
+		try:
+			return sum_loss / weighted_examples
+
+		except ZeroDivisionError:
+			return 0.
+
+
+	#------------------------------------------------------------------------------------------------		
+	def _calculate_since_last(self, sum_loss_since_last, weighted_examples_since_last):
+		"""Calculate since last 
+
+		Parameters
+		----------
+
+		sum_loss_since_last          : float
+										This is the difference between current iteration's sum_loss and previous iteration's sum_loss which is used to calculate since_last
+		weighted_examples_since_last : float
+										This is the difference between current iteration's weighted_examples and previous iteration's weighted_examples which is used to calculate since_last
+
+		Returns
+		-------
+
+		since_last      : float
+							since_last calculated using sum_loss_since_last and weighted_examples_since_last or if ZeroDivisionError is encountered then 0. is returned 
+		""" 
+		try:
+			return sum_loss_since_last / weighted_examples_since_last
+
+		except ZeroDivisionError:
+			return 0.
```

Comment:
I think these 2 methods could be removed since :
- they both do the same thing (guard against division by 0)
- they are only used once in the `fit ` method

Reply:
So should I instead write the code for these inside `.fit` method? At first I though that doing that would make the code look messy or less readable. What would you suggest on this?

Reply:
I agree with @etiennekintzler: The extra code in `.fit` is much smaller than the code + header of the helper functions, and what is going on is fairly straightforward. The only thing I would suggest is to do the zero check directly, rather than relying on the exception:

```
   return (sum_loss_since_last / weighted_examples_since_last) if weighted_examples_since_last  != 0 else 0
```